File: cpp/linklist_82.cpp
#include "linklist.h"
#include <iostream>

ListNode *deduplicate(ListNode *head) {
  ListNode dummy_head(0, head);
  ListNode *prev = &dummy_head;
  while (true) {
    ListNode *current = prev->next;
    if (current == nullptr) {
      return dummy_head.next;
    }

    ListNode *next = current->next;
    bool has_dup = false;
    while (next != nullptr && next->val == current->val) {
      has_dup = true;
      next = next->next;
    }
    if (has_dup) {
      prev->next = next;
    } else {
      prev = prev->next;
    }
    if (next == nullptr) {
      return dummy_head.next;
    }
  }
}

void testCase(const std::vector<int> &input) {
  ListNode *head = VecToList(input);
  head = deduplicate(head);
  while (head != nullptr) {
    std::cout << head->val << " ";
    head = head->next;
  }
  std::cout << std::endl;
}

int main() {
  testCase({1, 1, 2, 3, 3, 4, 6, 8, 8, 8, 9});
  testCase({1, 2, 3, 3, 4, 6, 8, 8, 8, 9, 9});
  testCase({1, 2, 3, 3, 4, 8, 8, 8, 9});
  testCase({1, 1, 2, 2, 3, 3, 3});
  testCase({1, 2, 3});
  return 0;
}

File: rust/src/linklist_82.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    let mut dummy_head = ListNode::new(0);
    dummy_head.next = head;
    let mut prev = &mut dummy_head;
    loop {
        if prev.next.is_none() {
            return dummy_head.next.take();
        }
        if prev.next.as_ref().unwrap().next.is_none() {
            return dummy_head.next.take();
        }
        while prev.next.as_ref().unwrap().val
            != prev.next.as_ref().unwrap().next.as_ref().unwrap().val
        {
            prev = prev.next.as_mut().unwrap();
            if prev.next.as_ref().unwrap().next.is_none() {
                return dummy_head.next.take();
            }
        }
        let mut del_end = prev.next.as_mut().unwrap().next.as_mut().unwrap();
        while del_end.next.is_some() && del_end.val == del_end.next.as_ref().unwrap().val {
            del_end = del_end.next.as_mut().unwrap();
        }
        prev.next = del_end.next.take();
    }
}
